osgp/shared/osgp-dto/src/main/java/org/opensmartgridplatform/dto/valueobjects/smartmetering/ConfigurationFlagTypeDto.java

```diff
@@ -9,17 +9,55 @@
  */
 package org.opensmartgridplatform.dto.valueobjects.smartmetering;
 
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+
 public enum ConfigurationFlagTypeDto {
 
-    DISCOVER_ON_OPEN_COVER,
-    DISCOVER_ON_POWER_ON,
-    DYNAMIC_MBUS_ADDRESS,
-    PO_ENABLE,
-    HLS_3_ON_P_3_ENABLE,
-    HLS_4_ON_P_3_ENABLE,
-    HLS_5_ON_P_3_ENABLE,
-    HLS_3_ON_PO_ENABLE,
-    HLS_4_ON_PO_ENABLE,
-    HLS_5_ON_PO_ENABLE;
+    DISCOVER_ON_OPEN_COVER(0, null, false),
+    DISCOVER_ON_POWER_ON(1, null, false),
+    DYNAMIC_MBUS_ADDRESS(2, null, false),
+    PO_ENABLE(3, 3, false),
+    HLS_3_ON_P_3_ENABLE(4, null, false),
+    HLS_4_ON_P_3_ENABLE(5, null, false),
+    HLS_5_ON_P_3_ENABLE(6, 6, false),
+    HLS_3_ON_PO_ENABLE(7, null, false),
+    HLS_4_ON_PO_ENABLE(8, null, false),
+    HLS_5_ON_PO_ENABLE(9, 9, false),
+    DIRECT_ATTACH_AT_POWER_ON(null, 10, false),
+    HLS_6_ON_P3_ENABLE(null, 11, false),
+    HLS_7_ON_P3_ENABLE(null, 12, false),
+    HLS_6_ON_P0_ENABLE(null, 13, false),
+    HLS_7_ON_P0_ENABLE(null, 14, false);
+
+    private final Integer bitPositionDsmr4;
+    private final Integer bitPositionSmr5;
+    private final boolean readOnly;
+
+    ConfigurationFlagTypeDto(final Integer bitPositionDsmr4, final Integer bitPositionSmr5, final boolean readOnly) {
```

Comment:
this seems like introducing a carthesian product into an object. the enum set seems to be mutually exclusive, consider making two enumsets. This will simplify the get logic.

Reply:
The sets of enums are not mutually exclusive, there are three elements they have in common.

We cannot avoid this complexity. We need to record the bit positions for each protocol. There will not be less values if the enum sets are split up, in fact there will be more.

Note that the readOnly value is always false. I left that in the code just in case the requirement to make certain flags read-only comes back.

Reply:
agreed.